### PR TITLE
Port analysis module to pytest

### PR DESCRIPTION
**Pull request by utkbansal**
Fixes #

Changes made in this Pull Request:
 - 

PR Checklist
------------
 - [ ] Tests?
 - [ ] Docs?
 - [ ] CHANGELOG updated?
 - [ ] Issue raised/referenced?

### CODE REVIEW

**Comment by utkbansal** [comment]
There is some unusual stuff going on in the `TestHydrogenBondAnalysisChecking` class. Can someone explain?

**Comment by richardjgowers** [comment]
@utkbansal not sure, but it might be because of `self.universe` being shared?  Try replacing `_setUp` with a fixture, and make the `_run` method not operate on `self` variables

**Comment by kain88-de** [comment]
@utkbansal looks like you have to replace `assertRaises` and `assert_warns` with pytest equivalents.

**Comment by utkbansal** [comment]
@richardjgowers @kain88-de What should I do with this setup?

```python
class TestEncoreImportWarnings(object):
    def setUp(self):
        # clear cache of encore module
        for mod in list(sys.modules):  # list as we're changing as we iterate
            if 'encore' in mod:
                sys.modules.pop(mod, None)

    @block_import('sklearn')
    def _check_sklearn_import_warns(self, package):
        warnings.simplefilter('always')
        assert_warns(ImportWarning, importlib.import_module, package)

    def test_import_warnings(self):
        for pkg in (
                'MDAnalysis.analysis.encore.dimensionality_reduction.DimensionalityReductionMethod',
                'MDAnalysis.analysis.encore.clustering.ClusteringMethod',
        ):
            yield self._check_sklearn_import_warns, pkg
```

This isn't something that a fixture is used for. This literally is some setup stuff. One option is to name this method `setup` and call it in every method. Any other non-hacky solutions?

**Comment by jbarnoud** [comment]
I did not test, but you should be able to do something like:

```python
@pytest.fixture()
def remove_encore_module():
    # clear cache of encore module
    for mod in list(sys.modules):  # list as we're changing as we iterate
        if 'encore' in mod:
            sys.modules.pop(mod, None)

@pytest.mark.usefixtures('remove_encore_module')
class TestEncoreImportWarnings(TestCase):
    ...
```

From what I understand, the `@pytest.mark.usefixtures` decorator make sure the fixture is called but do not gives you access to the result of the fixture directly. Since the fixture returns nothing (`None` actually), it does not matter that you cannot access it; but you do not polute your tests with a useless argument to call the fixture.

Once again, I did not test so I might be wrong.

**Comment by utkbansal** [comment]
`TestHydrogenBondAnalysisChecking` is ultra-confusing. Other than that, everything else works.

File: testsuite/MDAnalysisTests/analysis/test_encore.py [hunk 1]
```diff
@@ -826,20 +827,22 @@ def test_get_distance_matrix(self):
...
+    def test_import_warnings(self, recwarn):
+        for mod in list(sys.modules):  # list as we're changing as we iterate
+            if 'encore' in mod:
+                sys.modules.pop(mod, None)
         for pkg in (
                 'MDAnalysis.analysis.encore.dimensionality_reduction.DimensionalityReductionMethod',
                 'MDAnalysis.analysis.encore.clustering.ClusteringMethod',
         ):
-            yield self._check_sklearn_import_warns, pkg
+            self._check_sklearn_import_warns(pkg, recwarn)
```
Comment (jbarnoud) [inline testsuite/MDAnalysisTests/analysis/test_encore.py]: This changes the behaviour of the test. Instead of having 2 tests that can fail individually, there is only one test that has two consecutive asserts. It is fine for now, but please add a comment as it may be forgotten when you will move to proper pytest idioms.

**Comment by jbarnoud** [comment]
> TestHydrogenBondAnalysisChecking is ultra-confusing

It is confusing indeed, but what do you find confusing there?

**Comment by jbarnoud** [comment]
Please prioritize this PR so we can get rid of nose.

File: .travis.yml [hunk 2]
```diff
@@ -23,9 +23,9 @@ env:
     - PYTHON_VERSION=2.7
     - COVERALLS=false
     - NOSE_FLAGS="--processes=2 --process-timeout=400 --no-open-files --with-timer --timer-top-n 50"
-    - NOSE_TEST_LIST="analysis"
+    - NOSE_TEST_LIST=""
     - PYTEST_FLAGS="--disable-pytest-warnings -n 2"
-    - PYTEST_LIST="testsuite/MDAnalysisTests/lib testsuite/MDAnalysisTests/formats testsuite/MDAnalysisTests/coordinates testsuite/MDAnalysisTests/utils testsuite/MDAnalysisTests/topology testsuite/MDAnalysisTests/auxiliary testsuite/MDAnalysisTests/core"
+    - PYTEST_LIST="testsuite/MDAnalysisTests/lib testsuite/MDAnalysisTests/formats testsuite/MDAnalysisTests/coordinates testsuite/MDAnalysisTests/utils testsuite/MDAnalysisTests/topology testsuite/MDAnalysisTests/auxiliary testsuite/MDAnalysisTests/core testsuite/MDAnalysisTests/analysis"
```
Comment (orbeckst) [inline .travis.yml]: also remove `nose` from `package/setup.py` and `testsuite/setup.py` — yay, major step forward in your project: getting rid of `nose`.

File: .travis.yml [hunk 3]
```diff
@@ -23,9 +23,9 @@ env:
...
     - NOSE_COVERAGE_FILE="nose_coverage"
     - PYTEST_COVERAGE_FILE="pytest_coverage"
     - MAIN_CMD="pytest ${PYTEST_LIST} ${PYTEST_FLAGS}; python ./testsuite/MDAnalysisTests/mda_nosetests ${NOSE_TEST_LIST} ${NOSE_FLAGS}"
```
Comment (orbeckst) [inline .travis.yml]: remove  use of `mda_nosetests`

File: .travis.yml (hunk 2 above)
Comment (orbeckst) [inline .travis.yml]: And remove `testsuite/mda_nosetests.py`.

Comment (orbeckst) [inline .travis.yml]: @mnmelo @jbarnoud  @kain88-de @richardjgowers : should the nose-specific plugins also be removed at this time? We have to recreate the plugins for pytest or replace with native pytest plugins anyway,

File: testsuite/MDAnalysisTests/analysis/test_align.py [hunk 4]
```diff
@@ -43,8 +43,8 @@
 warnings.simplefilter('always')
 
 
-class TestRotationMatrix(object):
-    def __init__(self):
+class TestRotationMatrix(TestCase):
+    def setUp(self):
```
Comment (orbeckst) [inline testsuite/MDAnalysisTests/analysis/test_align.py]: @utkbansal I assume the goal of this PR is to make it run under pytest and then you will do a second round adapting it to pytest style? Is this correct? (Would be fine with me.)

File: testsuite/MDAnalysisTests/analysis/test_rms.py [hunk 5]
```diff
@@ -45,10 +48,10 @@
 warnings.simplefilter('always')
 
 
-class Testrmsd(object):
+class Testrmsd(TestCase):
     @dec.skipif(parser_not_found('DCD'),
```
Comment (orbeckst) [inline testsuite/MDAnalysisTests/analysis/test_rms.py]: remove DCD skipif, we have full DCD support now

File: testsuite/MDAnalysisTests/analysis/test_rms.py [hunk 6]
```diff
@@ -148,112 +150,118 @@ def test_with_superposition_equal(self):
...
                                   "test values")
 
-    @raises(SelectionError)
-    def test_ref_length_unequal_len(self):
+    def test_ref_length_unequal_len(self, universe):
         reference = MDAnalysis.Universe(PSF, DCD)
         reference.atoms = reference.atoms[:-1]
-        RMSD = MDAnalysis.analysis.rms.RMSD(self.universe,
-                                            reference=reference)
+        with pytest.raises(SelectionError):
```
Comment (orbeckst) [inline testsuite/MDAnalysisTests/analysis/test_rms.py]: good!

File: testsuite/MDAnalysisTests/analysis/test_hbonds.py [hunk 7]
```diff
@@ -239,7 +241,7 @@ def run_HBA_dynamic_selections(*args):
             self._tearDown()
 
 
-class TestHydrogenBondAnalysisTIP3P(object):
+class TestHydrogenBondAnalysisTIP3P(TestCase):
     @dec.skipif(parser_not_found('DCD'),
```
Comment (orbeckst) [inline testsuite/MDAnalysisTests/analysis/test_hbonds.py]: remove DCD skipif

File: .travis.yml (hunk 2 above)
Comment (kain88-de) [inline .travis.yml]: @orbeckst these changes should be in a separate PR. The plugins can pretty much all go. There is the openfile plugin but otherwise pytest comes with better alternatives

File: testsuite/MDAnalysisTests/analysis/test_align.py (hunk 4 above)
Comment (kain88-de) [inline testsuite/MDAnalysisTests/analysis/test_align.py]: yes that idea is correct

File: testsuite/MDAnalysisTests/analysis/test_rms.py (hunk 5 above)
Comment (kain88-de) [inline testsuite/MDAnalysisTests/analysis/test_rms.py]: better in a second PR

Comment (orbeckst) [inline testsuite/MDAnalysisTests/analysis/test_rms.py]: ok

File: testsuite/MDAnalysisTests/analysis/test_hbonds.py (hunk 7 above)
Comment (orbeckst) [inline testsuite/MDAnalysisTests/analysis/test_hbonds.py]: separate PR, as @kain88-de says

**Comment by kain88-de** [comment]
@utkbansal You should prioritize this PR. The other PR's are nice but not as important as this one. You still haven't told us your exact problem.

File: testsuite/MDAnalysisTests/analysis/test_hbonds.py [hunk 8]
```diff
@@ -207,9 +210,11 @@ def runOK():
                             raise AssertionError("HydrogenBondAnalysis protein/protein failed")
                         else:
                             return True
-                    yield runOK
+                    runOK()
```
Comment (jbarnoud) [inline testsuite/MDAnalysisTests/analysis/test_hbonds.py]: This change the behaviour and transforms a test generator into a succession of assertions in a single test. Please, add a comment so you can restore the initial behaviour latter.

File: testsuite/MDAnalysisTests/analysis/test_hbonds.py [hunk 9]
```diff
@@ -207,9 +210,11 @@ def runOK():
...
                 else:
-                    yield assert_raises, SelectionError, run_HBA, s1, s2, s1type
+                    # yield assert_raises, SelectionError, run_HBA, s1, s2, s1type
+                    with pytest.raises(SelectionError):
+                        run_HBA(s1, s2, s1type)
```
Comment (jbarnoud) [inline testsuite/MDAnalysisTests/analysis/test_hbonds.py]: Same here. Add a comment.

File: testsuite/MDAnalysisTests/analysis/test_hbonds.py [hunk 10]
```diff
@@ -234,12 +239,12 @@ def run_HBA_dynamic_selections(*args):
                         raise AssertionError("HydrogenBondAnalysis with update=True failed")
                     else:
                         return True
-                yield run_HBA_dynamic_selections, s1, s2, s1type
+                run_HBA_dynamic_selections(s1, s2, s1type)
```
Comment (jbarnoud) [inline testsuite/MDAnalysisTests/analysis/test_hbonds.py]: Idem. Comment.

File: testsuite/MDAnalysisTests/analysis/test_hbonds.py (hunk 8 above)
Comment (utkbansal) [inline testsuite/MDAnalysisTests/analysis/test_hbonds.py]: I am still working on it. This isn't final.

**Comment by utkbansal** [comment]
@kain88-de @richardjgowers  @jbarnoud This is the best hack I could come up with. Review please.

File: testsuite/MDAnalysisTests/analysis/test_hbonds.py [hunk 11]
```diff
@@ -162,6 +162,28 @@ def setUp(self):
...
 
 
+def build_parameters():
+    protein = "protein"
+    nothing = "resname ALA and not backbone"
+    params = []
+    for s1, s2, s1type in itertools.product((protein, nothing),
+                                            (protein, nothing),
+                                            ("donor", "acceptor", "both")):
+        params.append((s1, s2, s1type))
```
Comment (mnmelo) [inline testsuite/MDAnalysisTests/analysis/test_hbonds.py]: Must this become a list? Won't `pytest.mark.parametrize` accept the `itertools` generator object?

File: testsuite/MDAnalysisTests/analysis/test_hbonds.py [hunk 12]
```diff
@@ -162,6 +162,28 @@ def setUp(self):
...
+    return params
+
+
+def build_parameters2():
```
Comment (mnmelo) [inline testsuite/MDAnalysisTests/analysis/test_hbonds.py]: I'm not getting the rationale behind the two `build_parameters`. Is it to have the option of eventually supplying different lists of test args to the two tests?

File: testsuite/MDAnalysisTests/analysis/test_hbonds.py (hunk 11 above)
Comment (kain88-de) [inline testsuite/MDAnalysisTests/analysis/test_hbonds.py]: even if this has to be a list. Why not `list(itertools.product(...))`

File: testsuite/MDAnalysisTests/analysis/test_hbonds.py [hunk 13]
```diff
@@ -162,6 +162,28 @@ def setUp(self):
         self.values['acceptor_resnm'] = np.array([], dtype="<U3")
 
 
+def build_parameters():
```
Comment (mnmelo) [inline testsuite/MDAnalysisTests/analysis/test_hbonds.py]: Could you instead define `protein`, `nothing` and the `itertools.product` generator as class attributes of  `TestHydrogenBondAnalysisChecking`? This way the the `itertools.product` will be available when the decorator is used, and there won't be any need for the `build_parameters` functions.

File: testsuite/MDAnalysisTests/analysis/test_hbonds.py (hunk 11 above)
Comment (kain88-de) [inline testsuite/MDAnalysisTests/analysis/test_hbonds.py]: The iterator still works just fine.

```python
import pytest
import itertools

@pytest.mark.parametrize('a, b', itertools.product(range(3), range(3)))
def test_d(a, b):
    assert True
```

File: testsuite/MDAnalysisTests/analysis/test_hbonds.py (hunk 13 above)
Comment (utkbansal) [inline testsuite/MDAnalysisTests/analysis/test_hbonds.py]: Can't access `self` in `@pytest.mark.parametrize`

Comment (utkbansal) [inline testsuite/MDAnalysisTests/analysis/test_hbonds.py]: I've already removed` build_parameters()`

Comment (mnmelo) [inline testsuite/MDAnalysisTests/analysis/test_hbonds.py]: Class attrs don't need to be accessed via `self ` when in the class body. They're already in the namespace. Anyway, I think the solution of having the iterator call within the decorator is the best, as it clearer what's happening and allows customization in place.

Comment (utkbansal) [inline testsuite/MDAnalysisTests/analysis/test_hbonds.py]: Do you mean something like this would work?

```
class Foo(object):
    a = 2

    def bar(self):
        print a

f = Foo()
f.bar()
```

Comment (utkbansal) [inline testsuite/MDAnalysisTests/analysis/test_hbonds.py]: So is the latest change ok?

I've done this 
```

@pytest.mark.parametrize('s1, s2, s1type', itertools.product(
        ("protein", "resname ALA and not backbone"),
        ("protein", "resname ALA and not backbone"),
        ("donor", "acceptor", "both"),
    ))
```

**Comment by utkbansal** [comment]
`nose` is messing up now. It's trying to run everything.

**Comment by utkbansal** [comment]
@kain88-de @richardjgowers @jbarnoud The build technically passes because there are no errors in the pytest part.

**Comment by utkbansal** [comment]
@richardjgowers So would be able to merge this as is(once everyone confirms the changes)? I'm already working on #1480 to remove nose specific stuff.

File: testsuite/MDAnalysisTests/analysis/test_encore.py (hunk 1 above)
Comment (jbarnoud) [inline testsuite/MDAnalysisTests/analysis/test_encore.py]: The comment is still missing here.

File: testsuite/MDAnalysisTests/analysis/test_hbonds.py [hunk 14]
```diff
@@ -183,65 +184,74 @@ def _run(self, **kwargs):
...
-                            h = run_HBA(s1, s2, s1type)
-                        except:
-                            raise AssertionError("HydrogenBondAnalysis protein/protein failed")
-                        else:
-                            return True
-                    yield runOK
+        protein = "protein"
+
+        if s1 == s2 == protein:
+            def runOK():
```
Comment (jbarnoud) [inline testsuite/MDAnalysisTests/analysis/test_hbonds.py]: The way the test is written now, what is the point of having the nested function?

File: testsuite/MDAnalysisTests/analysis/test_hbonds.py [hunk 15]
```diff
@@ -183,65 +184,74 @@ def _run(self, **kwargs):
...
+                    return True
+
+            runOK()
+        else:
+            with pytest.raises(SelectionError):
+                self.run_HBA(s1, s2, s1type)
+
+        self._tearDown()
+
+    def run_HBA2(self, s1, s2, s1type):
```
Comment (jbarnoud) [inline testsuite/MDAnalysisTests/analysis/test_hbonds.py]: I do not like having functions named `run_HBA` and `run_HBA2`. It was OK that they had the same name when they were nested because the parent function was giving the context. Now the names do not tell how the two functions differ from each other. `run_HBA_no_update` and `run_HBA_update` would be better names, other names may work as well.

File: .travis.yml (hunk 3 above)
Comment (jbarnoud) [inline .travis.yml]: I agree with @orbeckst. With `NOSE_TEST_LIST` being empty, the call to `mda_nosetests` becomes buggy and should be removed. Removing the dependencies to nose is out of scope, though.

File: testsuite/MDAnalysisTests/analysis/test_hbonds.py (hunk 14 above)
Comment (utkbansal) [inline testsuite/MDAnalysisTests/analysis/test_hbonds.py]: Nothing really. I can pull that out.

**Comment by utkbansal** [comment]
@jbarnoud If I understand correctly, I have to squash all my commits now and then have one more commit that removes `mda_nosetests `. In the end, one commit for porting `analysis` module and one for removing `mda_nosetests `

**Comment by jbarnoud** [comment]
All your commits until the one titled "Quick fix" are logical units. I would squash them but I do not mind having them separate. From "Quick fix" and latter, the commits are iterative fixes that should be squashed. The removal of `mda_nosetest` must be on a separate commit.

**Comment by utkbansal** [comment]
It's hard to believe, but the full build completed in 16.5 mins.

File: testsuite/MDAnalysisTests/analysis/test_encore.py (hunk 1 above)
Comment (jbarnoud) [inline testsuite/MDAnalysisTests/analysis/test_encore.py]: The comment is bellow. I missed it.

**Comment by utkbansal** [comment]
@jbarnoud All good?

**Comment by utkbansal** [comment]
The build passed! Plus a minor increase in coverage.

**Comment by kain88-de** [comment]
@orbeckst do you have any more comments? I would like to merge this.
